src/http/http2.h

```diff
@@ -0,0 +1,604 @@
+    const uint8_t* name,
+    size_t namelen,
+    const uint8_t* value,
+    size_t valuelen,
+    uint8_t flags,
+    void* user_data)
+  {
+    auto k = std::string(name, name + namelen);
+    auto v = std::string(value, value + valuelen);
+    LOG_TRACE_FMT("http2::on_header_callback_client: {}:{}", k, v);
+
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    if (k == http2::headers::STATUS)
+    {
+      // Note: handle errors
+      stream_data->status = HTTP_STATUS_OK;
+    }
+    else
+    {
+      stream_data->headers.emplace(k, v);
+    }
+
+    return 0;
+  }
+
+  static int on_data_callback(
+    nghttp2_session* session,
+    uint8_t flags,
+    StreamId stream_id,
+    const uint8_t* data,
+    size_t len,
+    void* user_data)
+  {
+    LOG_TRACE_FMT("http2::on_data_callback: {}", stream_id);
+
+    auto* stream_data = get_stream_data(session, stream_id);
+
+    stream_data->request_body.insert(
+      stream_data->request_body.end(), data, data + len);
+
+    return 0;
+  }
+
+  static int on_data_callback_client(
+    nghttp2_session* session,
+    uint8_t flags,
+    StreamId stream_id,
+    const uint8_t* data,
+    size_t len,
+    void* user_data)
+  {
+    LOG_TRACE_FMT("http2::on_data_callback_client: {}", stream_id);
+
+    auto* stream_data = get_stream_data(session, stream_id);
+
+    stream_data->response_body.insert(
+      stream_data->response_body.end(), data, data + len);
+
+    return 0;
+  }
+
+  static int on_stream_close_callback(
+    nghttp2_session* session,
+    StreamId stream_id,
+    uint32_t error_code,
+    void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_stream_close_callback: {}, {}", stream_id, error_code);
+
+    // Note: Stream should be closed correctly here
+
+    return 0;
+  }
+
+  static ssize_t on_data_source_read_length_callback(
+    nghttp2_session* session,
+    uint8_t frame_type,
+    int32_t stream_id,
+    int32_t session_remote_window_size,
+    int32_t stream_remote_window_size,
+    uint32_t remote_max_frame_size,
+    void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_data_source_read_length_callback: {}, {}, allowed [1, "
+      "min({},{},{})]",
+      stream_id,
+      max_data_read_size,
+      session_remote_window_size,
+      stream_remote_window_size,
+      remote_max_frame_size);
+
+    return max_data_read_size;
+  }
+
+  static int on_error_callback(
+    nghttp2_session* session,
+    int lib_error_code,
+    const char* msg,
+    size_t len,
+    void* user_data)
+  {
+    LOG_FAIL_FMT("HTTP/2 error: {}", std::string(msg, msg + len));
+    return 0;
+  }
+
+  class ServerSession : public Session
+  {
+  private:
+    http::RequestProcessor& proc;
+
+  public:
+    ServerSession(http::RequestProcessor& proc_, ccf::Endpoint& endpoint_) :
+      Session(endpoint_, false),
+      proc(proc_)
+    {
+      LOG_TRACE_FMT("Initialising HTTP2 Server Session");
+
+      // Note: Should be set by operator on node startup
```

Comment:
I think this should be added to `struct NetInterface`, which would make it operator-configurable too.